### PR TITLE
Initial ping handshake and continuous block sync (issue #39)

### PR DESCRIPTION
**Pull request by uwiger**
- ping must be manually initiated (`aec_sync:connect_peer(Uri)`)
- peers share other peers during ping, leading to autom. discovery
- initial handshake protocol inspired by Ethereum's
- gossip protocol used to broadcast new blocks to all peers

### CODE REVIEW

File: rebar.config [hunk 1]
```diff
@@ -5,7 +5,9 @@
         {rocksdb, "0.12.0"},
         {cowboy, "1.0.4", {git, "https://github.com/ninenines/cowboy.git", {tag, "1.0.4"}}},
         {jsx, {git, "https://github.com/talentdeficit/jsx.git", {tag, "2.8.0"}}},
-        {jesse, {git, "https://github.com/for-GET/jesse.git", {tag, "1.5.0-rc2"}}}
+        {jesse, {git, "https://github.com/for-GET/jesse.git", {tag, "1.5.0-rc2"}}},
+        {gproc, "0.6.1"},
+        {jobs, "0.7.0"}
```
Comment (lucafavatella) [inline rebar.config]: I am not sure why jobs is not in the rebar.lock.

File: apps/aecore/src/aec_chain.erl [hunk 2]
```diff
@@ -281,6 +284,36 @@ force_insert_headers(HeaderChain) ->
...
+    case file:read_file(File) of
+        {ok, Bin} ->
+            try binary_to_term(Bin) of
+                {{0,1}, Headers, Blocks} ->
+                    io:fwrite("Headers = ~p~n"
+                              "Blocks = ~p~n",
+                              [Headers, Blocks]),
+                    lists:foreach(
+                      fun(H) ->
+                              insert_header(H)
```
Comment (lucafavatella) [inline apps/aecore/src/aec_chain.erl]: May you please elaborate on why no match to `ok`? I see the the export make such that `Headers` is from genesis to top.

File: apps/aecore/src/aec_chain.erl [hunk 3]
```diff
@@ -281,6 +284,36 @@ force_insert_headers(HeaderChain) ->
...
+                      end, Headers),
+                    lists:foreach(
+                      fun(B) ->
+                              write_block(B)
```
Comment (lucafavatella) [inline apps/aecore/src/aec_chain.erl]: May you please elaborate on why no match to `ok`?

File: rebar.config (hunk 1 above)
Comment (uwiger) [reply rebar.config]: Not sure either

File: apps/aecore/src/aec_chain.erl (hunk 2 above)
Comment (uwiger) [reply apps/aecore/src/aec_chain.erl]: The import caused a crash trying to deserialize transactions, so pending a fix of that, it doesn't work.

File: apps/aecore/src/aec_chain.erl (hunk 3 above)
Comment (uwiger) [reply apps/aecore/src/aec_chain.erl]: No reason. It should match on ok.

**Comment by uwiger** [comment]
Yes, there need to be tests, but for them to be practical, a fast way of creating blockchains is needed. Running two nodes on the same machine, I once waited 25 minutes for a single block to be mined.

**Comment by uwiger** [comment]
The failure in the test run (job 659.1) seems unrelated to this PR:

```
  1) aec_mining_tests:mine_block_from_genesis_test_/0: Find first block after genesis (PoW module aec_pow_cuckoo)
     Failure/Error: {error,
                        {badmatch,{error,key_not_found}},
                        [{aec_mining_tests,
                             '-mine_block_from_genesis_test_/0-fun-4-',0,
                             [{file,
                                  "/home/travis/build/aeternity/epoch/_build/test/lib/aecore/test/aec_mining_tests.erl"},
                              {line,218}]}]}
     Output: 
```

File: config/swagger.yaml [hunk 4]
```diff
@@ -15,31 +15,26 @@ schemes:
...
+          name: body
+          description: Ping a peer
+          required: true
+          schema:
+            $ref: '#/definitions/Ping'
       responses:
         '200':
           description: Sample response
           schema:
             $ref: '#/definitions/Ping'
```
Comment (velzevur) [inline config/swagger.yaml]: Shouldn't we have a 404 error response, too?

Comment (uwiger) [reply config/swagger.yaml]: Indeed

File: apps/aecore/src/aec_sync.erl [hunk 5]
```diff
@@ -29,25 +45,143 @@
...
 
 start_link() ->
-    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+    ensure_tab(?RECV_POOL, [set, public, named_table]),
+    case gen_server:start_link({local, ?MODULE}, ?MODULE, [], []) of
+        {ok, Pid} ->
+            ets:give_away(?RECV_POOL, Pid, []),
+            {ok, Pid};
+        Other ->
+            ets:delete(?RECV_POOL),
```
Comment (velzevur) [inline apps/aecore/src/aec_sync.erl]: As discussed - this could pose a danger

File: apps/aecore/src/aec_sync.erl [hunk 6]
```diff
@@ -29,25 +45,142 @@
...
+    #{<<"genesis_hash">> => base64:encode(GHash),
+      <<"best_hash">>    => base64:encode(TopHash),
+      <<"difficulty">>   => Difficulty,
+      <<"source">>       => Source,
+      <<"share">>        => 32,  % TODO: make this configurable
+      <<"peers">>        => []}.
+
+%% Compare local and remote 'Ping' objects.
+%% 1. Compare genesis blocks. If they differ, return error
+%% 2. Compare best_hash. If they are the same, we're in sync
```
Comment (sennui) [inline apps/aecore/src/aec_sync.erl]: best hash - is it ethereum thing to name like that top hash?

Comment (uwiger) [reply apps/aecore/src/aec_sync.erl]: That was borrowed from ethereum, yes.

**Pull request merged**